src/_posts/2017-08-04-splittable-do-fn.md

```diff
@@ -0,0 +1,449 @@
+[`UnboundedSource`]({{ site.baseurl }}/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/io/UnboundedSource.html)
+supports reporting the source's watermark and backlog.
+
+Unfortunately, these features come at a price. Coding against the Source API
+involves a lot of boilerplate and is error-prone, and it does not compose well
+with the rest of the Beam model, because a `Source` can appear only at the root
+of a pipeline. For example:
+
+* Using the Source API, it is not possible to read a `PCollection` of
+filepatterns;
+
+* A `Source` can not read a side input, or wait on another pipeline step to
+produce the data;
+
+* A `Source` can not emit an additional output (e.g. records that failed to
+parse) and so on.
+
+The Source API is not composable even with itself. For example, suppose Alice
+implemented an unbounded `Source` that watches a directory for new matching
+files, and Bob implemented an unbounded `Source` that tails a file. The Source
+API does not let them simply chain the sources together and obtain a `Source`
+that returns new records in new log files in a directory (a very common user
+request). Instead, such a source would have to be developed mostly from scratch,
+and our experience shows that a full-featured monolithic implementation of such
+a `Source` is incredibly difficult and error-prone.
+
+About 2 years ago we began thinking about how to address the limitations of the
+Source API, and ended up, surprisingly, addressing the limitations of `DoFn`
+instead.
+
+## Enter Splittable DoFn
+
+[Splittable DoFn](http://s.apache.org/splittable-do-fn) (SDF) is a
+generalization of `DoFn` that gives it the core capabilities of `Source` while
+retaining the `DoFn`s syntax, flexibility, modularity and ease of coding.  As a
+result, it becomes possible to develop more powerful IO connectors than before,
+with shorter, simpler, more reusable code.
+
+Note that, unlike `Source`, SDF *does not* have distinct bounded/unbounded APIs,
+just as regular `DoFn`s don't: there is only one API, which covers both of these
+use cases and anything in between. Thus, SDF closes the final gap in the unified
+batch/streaming programming model of Apache Beam.
+
+When reading the explanation of SDF below, keep in mind the running example of a
+`DoFn` that takes a filename as input and outputs the records in that file.
+People familiar with the `Source` API may find it useful to think of SDF as a
+way to read a `PCollection` of sources, treating the source itself as just
+another piece of data in the pipeline (*this, in fact, was one of the early
+design iterations among the work that lead to creation of SDF*).
+
+The two aspects where `Source` has an advantage over a regular `DoFn` are:
+
+* **Splittability:** applying a `DoFn` to a single element is *monolithic*, but
+reading from a `Source` is *non-monolithic*: the whole `Source` doesn't have to
+be read at once; rather, it is read in parts, called *bundles*. E.g. a large
```

Comment:
I find the use of the term "bundle" here a bit confusing--typically "bundles" refer to sets of elements, not parts of elements, and the key feature of bundles is the unit of commitment. Keeping with the current terminology, the parts are called restrictions. (It would be OK to say an element may now be spread/split across multiple bundles, i.e. the bundle  boundaries no longer need to be aligned with element boundaries).